docs/containers/balance-calculator/orchestration/index.md

```diff
@@ -0,0 +1,100 @@
+
+The Orchestrator is responsible for ensuring that all Accrual instances have been persisted which is a result of a successful application of the logical steps outlined below. 
+
+- Successful processing - if all of the logical steps outlined below succeed then the Orchestrator must tell the TimeCard events consumer to release the TimeCard event from the topic. 
+
+- Failed processing - if any of the logical steps outlined below fails then the Orchestrator must tell the TimeCard events consumer to hold on to the TimeCard event from the topic. At this point the Orchestrator should try to reprocesses the event. Note that the process outlined below as a series of logical steps plus the algorithm that is referenced below are designed to be idempotent. Therefore simply reprocessing the event until success can be done indefinitely
+
+Since the process outlined above has a dependency on the `callisto-accrual-restapi` it is possible that there could be a persistent failure to complete the process.
+
+The implementation should guard against this by using a combination of an [exponential back off retry for transitive errors and a circuit breaker for handling persistent failure](https://dzone.com/articles/understanding-retry-pattern-with-exponential-back)
+
+The algorithm for determining which Accrual type a TimeCard event should effect varies by Accrual type. On this basis it is envisaged that there will be a series of concrete implementations of an Accrual Type identifier interface and the Orchestrator simply cycles through each asking if the TimeCard event ties to the given Accrual type that the implementation knows about.
+
+More detail on the specifics of identifying each type of Accrual can be found in [accrual-type-identification.md](./accrual-type-identification.md)
+
+#### Accrual Finder
+Used to find the Accrual instances that are to be updated based on data in the TimeCard event and also the type(s) of Accrual that the Accrual Type identifier returned. More information on how to use TimeCard event data along with Accrual type data can be found in [accrual-balance-calculation.md](./accrual-balance-calculation.md). 
+
+Accrual instances are found via a RESTful call to the `callisto-accrual-restapi` component (`GET` [/resources/accruals](./../../rest-api/rest-endpoints.md#findaccruals))
+
+#### Balance calculator
+Having found the Accrual instances which are to be updated this component is responsible for calculating new balances and updating the owning Accrual instance. More information on how to use TimeCard event data to calculate a balance can be found in [accrual-balance-calculation.md](./accrual-balance-calculation.md).
+
+### Atomic transactions
+All Accruals resources that are passed to the `accuals-rest-api`'s batch update endpoint should be committed in a single transaction.
+
+In the design presented here all Accrual resources are sent in a single call to the batch update endpoint. 
```

Comment:
Is there any reason for the design to be agnostic regarding whether the accruals updates for all types are all committed in parallel or not? Could the design make it clear that only accruals of the same type need to be committed in a transaction?

If I’m reading right, it seems like the design leaves this up to engineers, but as it reaches between containers, to me it feels like something on which it’s worth giving architectural steer

Reply:
We didn't want to optimise prematurely. At the moment I don't think we really have a good feel for whether or not one large update is "better" than several smaller updates. The simplest thing for now is to perform one large update hence that approach being the one followed by the design. 

I think we might need to revisit that decision later but the RESTful API should not need to change. Instead we'd just end up calling it more times as part of calculating the accrual balances for a given person. Granted the internals of the balance calculator would have to change to chunk the work up but as I say, at the moment this feels like a premature optimisation